openedx_authz/rest_api/utils.py

```diff
@@ -0,0 +1,111 @@
+"""Utility functions for the Open edX AuthZ REST API."""
+
+from django.contrib.auth import get_user_model
+from django.db.models import Q
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from edx_rest_framework_extensions.auth.session.authentication import SessionAuthenticationAllowInactiveUser
+from rest_framework.permissions import IsAuthenticated
+
+from openedx_authz.rest_api.enums import SearchField, SortField, SortOrder
+
+User = get_user_model()
+
+
+def view_auth_classes(is_authenticated=True):
+    """
+    Function and class decorator that abstracts the authentication and permission checks for api views.
+    """
+
+    def _decorator(func_or_class):
+        """
+        Requires either OAuth2 or Session-based authentication.
+        """
+        func_or_class.authentication_classes = [
+            JwtAuthentication,
+            SessionAuthenticationAllowInactiveUser,
+        ]
+        if is_authenticated:
+            func_or_class.permission_classes.insert(0, IsAuthenticated)
+        return func_or_class
+
+    return _decorator
+
+
+def get_user_by_username_or_email(username_or_email: str) -> User:
+    """
+    Retrieve a user by their username or email address.
+
+    Args:
+        username_or_email (str): The username or email address to search for.
+
+    Returns:
+        User: The User object if found and not retired.
+
+    Raises:
+        User.DoesNotExist: If no user matches the provided username or email,
+            or if the user has an associated retirement request.
+    """
+    user = User.objects.get(Q(email=username_or_email) | Q(username=username_or_email))
+    if hasattr(user, "userretirementrequest"):
+        raise User.DoesNotExist
+    return user
+
+
+def sort_users(
+    users: list[dict], sort_by: SortField = SortField.USERNAME, order: SortOrder = SortOrder.ASC
+) -> list[dict]:
+    """
+    Sort users by a given field and order.
+
+    Args:
+        users (list[dict]): The users to sort.
+        sort_by (SortField, optional): The field to sort by. Defaults to SortField.USERNAME.
+        order (SortOrder, optional): The order to sort by. Defaults to SortOrder.ASC.
+
+    Raises:
+        ValueError: If the sort field is invalid.
+        ValueError: If the sort order is invalid.
+
+    Returns:
+        list[dict]: The sorted users.
+    """
+    if sort_by not in SortField.values():
+        raise ValueError(f"Invalid field: '{sort_by}'. Must be one of {SortField.values()}")
+
+    if order not in SortOrder.values():
+        raise ValueError(f"Invalid order: '{order}'. Must be one of {SortOrder.values()}")
+
+    sorted_users = sorted(users, key=lambda user: (user.get(sort_by) or "").lower(), reverse=order == SortOrder.DESC)
+    return sorted_users
+
+
+def filter_users(users: list[dict], search: str | None, roles: list[str] | None) -> list[dict]:
```

Comment:
We should keep an eye on the performance of this and the sort function, in case the data structures we use make them too slow. Can we run a simple load test in the sandbox environment?